File: backend/src/main/java/com/backend/api/review/service/AiReviewService.java
package com.backend.api.review.service;

import com.backend.api.question.service.AiQuestionService;
import com.backend.api.resume.service.ResumeService;
import com.backend.api.review.dto.request.AiReviewbackRequest;
import com.backend.api.review.dto.response.AiReviewResponse;
import com.backend.domain.resume.entity.Resume;
import com.backend.domain.review.entity.Review;
import com.backend.domain.review.repository.ReviewRepository;
import com.backend.domain.user.entity.User;
import com.backend.global.exception.ErrorCode;
import com.backend.global.exception.ErrorException;
import com.fasterxml.jackson.core.JsonProcessingException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class AiReviewService {

    private final ReviewRepository reviewRepository;
    private final ResumeService resumeService;
    private final AiQuestionService aiQuestionService;

    @Transactional
    public AiReviewResponse createAiReview(User user) throws JsonProcessingException {
        if (!user.isPremium()) {
            throw new ErrorException(ErrorCode.AI_FEEDBACK_FOR_PREMIUM_ONLY);
        }

        Resume resume = resumeService.getResumeByUser(user);
        AiReviewbackRequest request = AiReviewbackRequest.of(resume);
        String feedbackContent = aiQuestionService.getAiReviewContent(request);

        Review reviewEntity = Review.builder()
                .user(user)
                .resume(resume)
                .AiReviewContent(feedbackContent)
                .build();

        reviewRepository.save(reviewEntity);

        return AiReviewResponse.of(reviewEntity.getId(), feedbackContent, reviewEntity.getCreateDate());
    }

    public AiReviewResponse findReviewById(Long reviewId, User user) {
        Review review = reviewRepository.findById(reviewId)
                .orElseThrow(() -> new ErrorException(ErrorCode.NOT_FOUND_REVIEW));

        if (!Objects.equals(review.getUser().getId(), user.getId())) {
            throw new ErrorException(ErrorCode.ACCESS_DENIED_REVIEW);
        }

        return AiReviewResponse.of(review.getId(), review.getAiReviewContent(), review.getCreateDate());
    }

    public List<AiReviewResponse> findMyAiReviews(User user) {
        List<Review> reviews = reviewRepository.findByUserIdOrderByIdDesc(user.getId());

        return reviews.stream()
                .map(review -> AiReviewResponse.of(review.getId(), review.getAiReviewContent(), review.getCreateDate()))
                .collect(Collectors.toList());
    }
}

File: backend/src/main/java/com/backend/api/review/service/AiReviewService.kt
package com.backend.api.review.service

import com.backend.api.question.service.AiQuestionService
import com.backend.api.resume.service.ResumeService
import com.backend.api.review.dto.request.AiReviewbackRequest
import com.backend.api.review.dto.response.AiReviewResponse
import com.backend.domain.review.entity.Review
import com.backend.domain.review.repository.ReviewRepository
import com.backend.domain.user.entity.User
import com.backend.global.exception.ErrorCode
import com.backend.global.exception.ErrorException
import com.fasterxml.jackson.core.JsonProcessingException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class AiReviewService(
    private val reviewRepository: ReviewRepository,
    private val resumeService: ResumeService,
    private val aiQuestionService: AiQuestionService
) {

    @Transactional
    @Throws(JsonProcessingException::class)
    fun createAiReview(user: User?): AiReviewResponse {
        if (user == null) {
            throw ErrorException(ErrorCode.UNAUTHORIZED_USER)
        }

        if (!user.isPremium()) {
            throw ErrorException(ErrorCode.AI_FEEDBACK_FOR_PREMIUM_ONLY)
        }

        val resume = resumeService.getResumeByUser(user)
        val request = AiReviewbackRequest.of(resume)
        val feedbackContent = aiQuestionService.getAiReviewContent(request)

        val reviewEntity = Review.builder()
            .user(user)
            .resume(resume)
            .AiReviewContent(feedbackContent)
            .build()

        reviewRepository.save(reviewEntity)


        return AiReviewResponse.of(reviewEntity.id, feedbackContent, reviewEntity.createDate)
    }

    fun findReviewById(reviewId: Long, user: User?): AiReviewResponse {
        val review = reviewRepository.findById(reviewId)
            .orElseThrow { ErrorException(ErrorCode.NOT_FOUND_REVIEW) }

        if (user == null || review.user?.id != user.id) {
            throw ErrorException(ErrorCode.ACCESS_DENIED_REVIEW)
        }

        return AiReviewResponse.of(review.id, review.AiReviewContent, review.createDate)
    }

    fun findMyAiReviews(user: User?): List<AiReviewResponse> {
        if (user == null) {
            throw ErrorException(ErrorCode.UNAUTHORIZED_USER)
        }


        val reviews: List<Review> = reviewRepository.findByUserIdOrderByIdDesc(user.id)


        return reviews.stream()
            .map { review ->
                AiReviewResponse.of(
                    review.id,
                    review.AiReviewContent,
                    review.createDate
                )
            }
            .toList()
    }
}